File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> umap;
        stack<int> st;

        int n = nums2.size();

        for (int i = n - 1; i >= 0; i--) {

            int num = nums2[i];

            while (!st.empty() && st.top() <= num) {
                st.pop();
            }

            int greaterVal = st.empty() ? -1 : st.top();
            umap[num] = greaterVal;

            st.push(num);
        }

        vector<int> result;

        for (auto it : nums1) {
            result.push_back(umap[it]);
        }
        return result;
    }
};

File: 496-next-greater-element-i/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {

        int n = nums2.length;
        HashMap<Integer, Integer> umap = new HashMap<>();
        Stack<Integer> st = new Stack<>();

        for (int i = n - 1; i >= 0; i--) {
            int ch = nums2[i];
            while (!st.isEmpty() && st.peek() <= ch) {
                st.pop();
            }

            int value = st.empty() ? -1 : st.peek();

            umap.put(ch, value);
            st.push(ch);
        }

        int[] answer = new int[nums1.length];

        for (int i = 0; i < nums1.length; i++) {
            answer[i] = umap.get(nums1[i]);
        }
        return answer;
    }
}